core/trino-main/src/main/java/io/trino/cost/TopNStatsRule.java

```diff
@@ -47,8 +47,13 @@ protected Optional<PlanNodeStatsEstimate> doCalculate(TopNNode node, StatsProvid
         PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
         double rowCount = sourceStats.getOutputRowCount();
 
-        if (node.getStep() != TopNNode.Step.SINGLE) {
-            return Optional.empty();
+        /* CreatePartialTopN rule runs after ReorderJoins but before DetermineJoinDistributionType and DetermineSemiJoinDistributionType.
+         * Therefore, it is useful to provide estimates for partial and final topN nodes.
+         * The overall output of partial topN will depend on the number of splits that get created at runtime, and that's not known in advance.
+         * We use the conservative estimate of sourceStats for partial topN so that final topN can be accurately estimated.
+         */
+        if (node.getStep() == TopNNode.Step.PARTIAL) {
+            return Optional.of(sourceStats);
```

Comment:
We could assume some minimal reasonable split size.
For example, "TOP 5" is going to reduce data heavily (except for a pretty uncommon situation where splits return up to 5 rows), and "TOP 1000000" is not going to reduce data much (splits are often less than million rows).

BTW, the comment here seems to assume partial TopN is applied directly in the source stage, which may or may not be the same. 
Does it matter? Can we improve the comment?

Reply:
I've updated the comment here to clarify that we're talking about an example scenario, not implying that partialTopN can only be in source stage.
For HASH stages maybe it could be estimated as `(input / (hash partitions count * task concurrency)) * N`, but I don't think we know the stage partitioning at this point.
I've updated the partial topN estimate calculation to assume 1 million rows input.